src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs

```diff
@@ -0,0 +1,211 @@
+            private bool _itemAvailable = false;
+            /// <summary>Whether <see cref="MoveNextAsync"/> should use <see cref="_promise"/> for its return value.</summary>
+            private bool _usePromiseForResult = false;
+            /// <summary>The awaiter for WaitToReadAsync calls.</summary>
+            private ConfiguredValueTaskAwaitable<bool>.ConfiguredValueTaskAwaiter _waitToReadAwaiter;
+
+            /// <summary>The current item to be returned from <see cref="IAsyncEnumerator{T}.Current"/>.</summary>
+            private T _current;
+
+            private enum State
+            {
+                NotEnumerated = -1,
+                OuterLoop = 0,
+                FinishingAwait = 1,
+                HaveWaitToReadResult = 2,
+                TryRead = 3,
+                Completing = 4,
+                Done = -2
+            }
+
+            internal AsyncEnumerable(ChannelReader<T> reader)
+            {
+                Debug.Assert(reader != null);
+
+                _reader = reader;
+                _state = (int)State.NotEnumerated;
+                _builder = AsyncIteratorMethodBuilder.Create();
+            }
+
+            public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken)
+            {
+                // Use this instance if it's never been enumerated before; otherwise, since it might still be in use,
+                // create a new instance.
+                AsyncEnumerable inst = Interlocked.CompareExchange(ref _state, (int)State.OuterLoop, (int)State.NotEnumerated) == (int)State.NotEnumerated ?
+                    this :
+                    new AsyncEnumerable(_reader) { _state = (int)State.OuterLoop };
+                inst._cancellationToken = cancellationToken;
+                return inst;
+            }
+
+            public ValueTask<bool> MoveNextAsync()
+            {
+                // Reset state for the next call.
+                _promise.Reset();
+                _usePromiseForResult = false;
+                _itemAvailable = false;
+
+                // Push the state machine forward.
+                MoveNext();
+
+                // If the operation completed asynchronously, return a ValueTask wrapping our promise-based implementation.
+                // If the operation completed synchronously, return a bool-backed ValueTask.
+                return _usePromiseForResult ?
+                    new ValueTask<bool>(this, _promise.Version) :
+                    new ValueTask<bool>(_itemAvailable);
+            }
+
+            public void MoveNext()
+            {
+                try
+                {
+                    bool waitToReadResult = false;
+                    switch ((State)_state)
+                    {
+                        case State.OuterLoop:
+                            ValueTask<bool> waitToReadTask = _reader.WaitToReadAsync(_cancellationToken);
```

Comment:
>_reader.WaitToReadAsync(_cancellationToken) [](start = 61, length = 43)

do we need to have .ConfigureAwait(false) here?

Reply:
No, it's below when actually creating the awaiter.